src/util/metrics.ts

```diff
@@ -26,32 +118,245 @@ export default class Metrics_App {
+
+			const extrinsics_per_second = this.metrics['sas_extrinsics_per_second'] as client.Histogram;
+			const seconds = end();
+
+			extrinsics_per_second
+				.labels({ method: req.method, route: this.getRoute(req), status_code: res.statusCode })
+				.observe(extrinscs / seconds);
+
+			const extrinsics_per_block = this.metrics['sas_extrinsics_per_block'] as client.Histogram;
+
+			const blocks = 1;
+			extrinsics_per_block
+				.labels({ method: req.method, route: this.getRoute(req), status_code: res.statusCode })
+				.observe(extrinscs / blocks);
+
+			const seconds_per_block = this.metrics['sas_seconds_per_block'] as client.Histogram;
+			seconds_per_block
+				.labels({ method: req.method, route: this.getRoute(req), status_code: res.statusCode })
+				.observe(seconds / blocks);
+		}
+
+		if (req.query?.range) {
+			let totExtrinsics = 0;
+			if (Array.isArray(body)) {
+				totExtrinsics = body.reduce((current: number, block: { [x: string]: unknown }) => {
+					const extrinsics = block['extrinsics'];
+
+					if (Array.isArray(extrinsics)) {
+						return current + extrinsics.length;
+					}
+
+					return current;
+				}, 0);
+			} else {
+				const extrinsics = body['extrinsics'];
+				if (Array.isArray(extrinsics)) {
+					totExtrinsics = extrinsics.length;
+				}
+			}
+
+			const extrinsics_in_request = this.metrics['sas_extrinsics_in_request'] as client.Histogram;
+			extrinsics_in_request
+				.labels({ method: req.method, route: this.getRoute(req), status_code: res.statusCode })
+				.observe(totExtrinsics);
+
+			const extrinsics_per_second = this.metrics['sas_extrinsics_per_second'] as client.Histogram;
+			const seconds = end();
+			extrinsics_per_second
+				.labels({ method: req.method, route: this.getRoute(req), status_code: res.statusCode })
+				.observe(totExtrinsics / seconds);
+
+			const extrinsics_per_block = this.metrics['sas_extrinsics_per_block'] as client.Histogram;
+			const blocks = Array.isArray(body) ? body.length : 1;
+			extrinsics_per_block
+				.labels({ method: req.method, route: this.getRoute(req), status_code: res.statusCode })
+				.observe(totExtrinsics / blocks);
+
+			const seconds_per_block = this.metrics['sas_seconds_per_block'] as client.Histogram;
+			seconds_per_block
+				.labels({ method: req.method, route: this.getRoute(req), status_code: res.statusCode })
+				.observe(seconds / blocks);
+		}
+	}
+
+	preMiddleware() {
+		return (req: Query, res: Response, next: () => void) => {
+			const tot_requests = this.metrics['sas_total_requests'] as client.Counter;
+
+			// request count metrics
+			tot_requests.inc();
+			const request_duration_seconds = this.metrics['sas_request_duration_seconds'] as client.Histogram;
+			const end = request_duration_seconds.startTimer();
+
+			const oldJson = res.json;
+			res.json = (body: Body) => {
+				res.locals.body = body;
+				return oldJson.call(res, body);
+			};
+
+			res.once('finish', () => {
+				if (res.statusCode >= 400) {
```

Comment:
```suggestion
				if (res.statusCode >= 400 && req.originalUrl != '/favicon.ico') {
```
I am suggesting this because with each request/endpoint we query, a second request is done for the favicon on the background and we get a `warn: GET /favicon.ico 404 3ms` in the console which comes by default from `pjs-api`. 

With this addition 
- we ignore the `favicon` request and do not count it as an error.
- the `sas_request_errors` counter will only increase if there is an actual error in the endpoint requested.
- it is less accurate because there is a GET request that is still being made but it depends on what you actually want to count with the error counter.

Without this addition
- with every successful request we will see both `sas_request_success` counter and `sas_request_errors` counter increase (due to the erroneous favicon request) which is a little bit confusing.
- even though this is more accurate as number of requests and errors, again it depends on what you want to count with these metrics.

Mentioning this just to highlight what is happening here but it is up to you on how you will move forward.